SA: use wrapped database maps/transactions.

cpu opened the pull request:
New types and related infrastructure are added to the `db` package to allow wrapping gorp DbMaps and Transactions.

The wrapped versions return a special `db.ErrDatabaseOp` error type when errors occur. The new error type includes additional information such as the operation that failed and the related table.

Where possible we determine the table based on the types of the gorp function arguments. Where that isn't possible (e.g. with raw SQL queries) we try to use a simple regexp approach to find the table name. This isn't great for general SQL but works well enough for Boulder's existing SQL queries.

To get additional confidence my regexps work for all of Boulder's queries I temporarily changed the `db` package's `tableFromQuery` function to panic if the table couldn't be determined. I re-ran the full unit and integration test suites with this configuration and saw no panics.

Resolves https://github.com/letsencrypt/boulder/issues/4559

jsha commented on db/map.go:
```diff
@@ -0,0 +1,319 @@
+
+var (
+	// selectTableRegexp matches the table name from an SQL select statement
+	selectTableRegexp = regexp.MustCompile(`(?i)^\s*select\s+[a-z\d\.\(\), \_\*` + "`" + `]+\s+from\s+([a-z\d\_,` + "`" + `]+)`)
```
Since these regexes need to use ``` ` ``` and don't need to use `"`, it'll probably be clearer to quote them with `"` and drop the concatenations.

cpu replied:
Won't that require double escaping lots of other things (e.g. `\s`?)

I was not thrilled to realize there's no way to put a backtick literal into a Golang raw string but using a regular string seemed like it caused more problems than it solved.

jsha replied:
Oh right, good point. I had missed that part.

jsha commented on db/map.go:
```diff
@@ -0,0 +1,319 @@
+	gorp.SqlExecutor
+}
+
+func (we WrappedExecutor) Get(holder interface{}, keys ...interface{}) (interface{}, error) {
```
It looks like Get, Update, Insert, and Delete don't call `tableFromQuery`. It seems like maybe that's because we usually have a type we can use instead of going through the regexps, but we do have regexps defined for insert and delete. Did you intend to add tableFromQuery calls to those functions?

cpu replied:
> Did you intend to add tableFromQuery calls to those functions?

No, but I wanted to catch the cases where someone `Exec`'s an insert or delete. Those do require going through `tableFromQuery`.

jsha replied:
Thanks for the explanation. I think this makes sense to explain in a comment somewhere.

jsha commented on db/map.go:
```diff
@@ -0,0 +1,319 @@
+// wrap error results in ErrDatabaseOp instances before returning them to the
+// caller.
+type WrappedExecutor struct {
+	gorp.SqlExecutor
```
This type definition embeds `gorp.SqlExecutor`, but all the wrapped functions explicitly reference `we.SqlExecutor`. So we have the option of making it a field rather than an embed. That might make it easier in turn to embed WrappedExecutor in WrappedTransaction and WrappedMap, since it would only offer the specific methods that we know we care about.

Or are there other methods that WrappedExecutor does not implement, that we want to pass through to the `gorp.SqlExecutor`?

cpu replied:
@jsha Changing to a field means `WrappedExecutor` no longer meets `gorp.SqlExecutor` which means it can't be returned from the `WithContext` functions of a transaction or a map, both of which need to return a `gorp.SqlExecutor`. If I change to using a field then `WrappedExecutor` needs to wrap `Query`, `SelectInt`, `SelectNullInt`, etc etc

jsha replied:
Makese sense.

jsha commented on db/map.go:
```diff
@@ -0,0 +1,319 @@
+
+func (we WrappedExecutor) Insert(list ...interface{}) error {
+	if err := we.SqlExecutor.Insert(list...); err != nil {
+		table := "unknown"
```
This chunk is small but repeated a few times and would be a good candidate for a small helper function.

jsha commented on db/map.go:
```diff
@@ -0,0 +1,319 @@
+	result, err := we.SqlExecutor.Select(holder, query, args...)
+	if err != nil {
+		table := fmt.Sprintf("%T", holder)
+		if extractedTable := tableFromQuery(query); extractedTable != "" {
```
Since we expect in most cases to be able to extract the table, I think it makes more sense to put this function (and similar ones) in order by expected flow:

```
table := tableFromQuery(query)
if table == "" {
  table := fmt.Sprintf("%T", holder)
}
```

rolandshoemaker replied:
Related: when we can't explicitly derive the table name we should probably indicate that in the error, i.e. `fmt.Sprintf("%T (unknown table)", holder)`.

jsha commented on db/mocks.go:
```diff
@@ -46,16 +46,13 @@ type DatabaseMap interface {
-	Delete(...interface{}) (int64, error)
-	Get(interface{}, ...interface{}) (interface{}, error)
-	Update(...interface{}) (int64, error)
+	gorp.SqlExecutor
```
`SqlExecutor` is a broader type than our current `Transaction`. For instance, it includes `SelectInt`, `SelectNullInt`, etc. Is there a reason to switch to gorp's interface type? FWIW I think it's good Go practice to define narrower interfaces over other packages' types, to make mocking easier.

jsha commented on sa/database.go:
```diff
@@ -8,14 +8,17 @@ import (
 	blog "github.com/letsencrypt/boulder/log"
 )
 
+// TODO(@cpu): Update NewDbMap godoc comment
```
Was this a TODO for this PR?

jsha commented on sa/database.go:
```diff
@@ -38,9 +41,11 @@ var setMaxOpenConns = func(db *sql.DB, maxOpenConns int) {
 	db.SetMaxOpenConns(maxOpenConns)
 }
 
+// TODO(@cpu): Update NewDbMapFromConfig godoc comment
```
Ditto

rolandshoemaker commented on db/map.go:
```diff
@@ -0,0 +1,319 @@
+
+func (we WrappedExecutor) Insert(list ...interface{}) error {
+	if err := we.SqlExecutor.Insert(list...); err != nil {
+		table := "unknown"
```
Instead of deriving the table from the type, we could alter the `initTables` method to record the actual type <-> table mapping somewhere and use that here.

cpu replied:
This is a fairly big shift from the existing code. I'd rather take it as a follow-up if folks feel strongly. It does sound like it may be slightly better UX.

rolandshoemaker replied:
:+1: works as a follow-up for me.

cpu replied:
Filed https://github.com/letsencrypt/boulder/issues/4594 so it doesn't get forgotten. Thanks!

cpu commented:
@rolandshoemaker @jsha Ready for another pass. I believe I've addressed all feedback except [this one comment](https://github.com/letsencrypt/boulder/pull/4585#discussion_r352810811) from @jsha I replied to.
